new module to check query buffer operations

hmmr opened the pull request:
RTS-733, RTS-870, RTS-1006

Depends on https://github.com/basho/riak_kv/pull/1479, which see.

hmmr committed:
new module to check query buffer operations

hmmr committed:
use {allow_qbuf_reuse, true} in query/5 Options, instead of ONLY keyword

hmmr committed:
ts_cluster_coverage: no MaxQuantum any more

hmmr committed:
query_buffers: milliseconds in timestamps

also, extend time range.

hmmr committed:
ts_simple_query_buffers: proofread and add more comments

hmmr committed:
ts_simple_query_buffers_ldb_errors: new module

hmmr committed:
query buffers: new tests for query_too_big and ldb errors

hmmr committed:
Merge branch 'riak_ts-develop' into feature-az-query_buffers

hmmr committed:
ts_simple_qbuf_ldb: reset writing perms on .../query_buffers dir

gordonguthrie commented on tests/ts_cluster_coverage.erl:
```diff
@@ -1,6 +1,6 @@
 % -------------------------------------------------------------------
```
Running this test thows crash reports

```
2016-10-16 08:52:05 =ERROR REPORT====
Error in process <0.155.0> on node 'riak_test@127.0.0.1' with exit value: {{nocatch,got_right_results_from_wrong_vnode},[{ts_cluster_coverage,make_noise_if_results_not_empty,5,[{file,"tests/ts_cluster_coverage.erl"},{line,183}]},{ts_cluster_coverage,'-test_quanta_range/5-fun-1-',3,[{file...
```

gordonguthrie commented on tests/ts_simple_query_buffers_SUITE.erl:
```diff
@@ -0,0 +1,314 @@
+                  [F || F <- FF, F /= undefined]
+          end,
+          OrderByBattery1),
+    lists:foreach(
```
Anonymous functions inside `list:foreach` are hard to decipher, doubling it makes it worse and makes the risk of unspotted unintended capture very, very high

hmmr replied:
This is carefully hand-crafted to generate a those test queries. It's definitely not the most readable code. To help understand its purpose, I'm specifically advising to uncomment the `ct:log` entries in `check_sorted` printing each query as we go (but then comment it again because of the huge volumes of logs it would generate).

gordonguthrie commented on tests/ts_simple_query_buffers_SUITE.erl:
```diff
@@ -0,0 +1,314 @@
+    %% including sorting according to ORDER BY
+    %% uncomment these log entries to inspect the data visually (else
+    %% it's just going to blow up the log into tens of megs):
+    %% ct:log("Query: \"~s\"", [Query]),
```
`ct:pal` is probably better here

gordonguthrie reviewed (commented):
We need to review the test suites again in light of getting a list of all error conditions. Also the test suite needs to extend from simple to cluster and tests need to be added to the upgrade/downgrade ladder.

Once again we need to think about writing the EQC SQL generator tests...

hmmr committed:
ts_simple_query_buffers: move error handling tests to main module

with a new test added to check for max query quanta error condition.

hmmr committed:
rt_client: correctly work with both stagedevrel and devrel

hmmr committed:
ts_simple_query_buffers: the default query_buffers path has changed

with some function renames

hmmr committed:
ts_updown_util: hooks to modify riak.conf for curr/prev versions

hmmr committed:
ts_updown_util: honour use_previous_client in CREATE, INSERT queries

hmmr committed:
whitespace cleanup

hmmr committed:
ts_updown_util: ct:log cluster node versions as they are before CREATE

hmmr committed:
rt_client: supply correct paths

generally, make the bloody thing work for the first time.

hmmr committed:
ts_updown_util: correctly emulate rt:pbc on previous client node

hmmr committed:
ts_cluster_updowngrade_group_by: expect all scenarios vs queries to pass

hmmr committed:
ts_updown_util: report cluster composition before CREATE and before SELECT

hmmr committed:
ts_updown_util: use wait_until_capability before issuing queries

This is to ensure that on node transition, any capabilities change
take effect across all nodes.

hmmr committed:
new module ts_cluster_updowngrade_order_by_SUITE

hmmr commented:
The commits above, along with the new module ts_cluster_updowngrade_order_by_SUITE, contain these fixes to the TS up/downgrade facilities:
1. Correctly set up and start the previous client slave node. Add `{use_previous_client, true}` to the proplist in your `Config`, which needs to be returned by `make_initial_config(Config)` to take effect (  4b61877).
2. Add a pair of hooks, to be called before a transitioned node is started, with the path to its riak.conf. Useful when the newer version introduces new keys which will need to be removed from the same riak.conf before the older riak version can use it. This was previously hardcoded for the case of 1.3 vs 1.4, in ts_cluster_updowngrade_group_by_SUITE, and is now generalized (229f809, 7707de0).
3. After starting a transitioned node, we need to give it some extra time for the changed capabilities to percolate in the cluster (e188bfa).

hmmr committed:
Merge branch 'riak_ts-develop' into feature-az-query_buffers

hmmr committed:
ts_updown_util: rework to halt scenario if a stage fails

hmmr committed:
ts_cluster_updowngrade_*: exclude downgrades from scenarios, for 1.5-1.4

hmmr committed:
ts_updown_util: more checkpoint messages in logs

hmmr committed:
ts_cluster_updowngrade_: reenable previous QueryNodeVsn (exc. order_by)

hmmr committed:
ts_updown_util: teach find_or_make_node_at_vsn to properly select nodes

hmmr committed:
ts_updown_util: more logging messages

hmmr committed:
ts_cluster_updowngrade_*: enable more scenarios, add new caps to check

this requires tag riak_ts-1.4.1 in riak_kv

gordonguthrie commented:
+1 eff8df9

hmmr commented:
@borshop merge
